diff --git a/external/nginx-rtmp-module/ngx_rtmp_eval.c b/external/nginx-rtmp-module/ngx_rtmp_eval.c
index bf4c86144..919dd32e2 100644
--- a/external/nginx-rtmp-module/ngx_rtmp_eval.c
+++ b/external/nginx-rtmp-module/ngx_rtmp_eval.c
@@ -89,7 +89,7 @@ ngx_rtmp_eval_append_var(void *ctx, ngx_buf_t *b, ngx_rtmp_eval_t **e,
     ngx_rtmp_eval_t    *ee;
 
     for (; *e; ++e) {
-        for (k = 0, ee = *e; ee->handler; ++k, ++ee) {
+        for (k = 0, ee = *e; IA2_ADDR(ee->handler); ++k, ++ee) {
             if (ee->name.len == name->len &&
                 ngx_memcmp(ee->name.data, name->data, name->len) == 0)
             {
-- 
